doc/adr/0005-wego-core-auth-strategy.md

```diff
@@ -0,0 +1,83 @@
+  - Notice that `localhost` redirect_uris are permitted: https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#localhost-redirect-urls
+  - This approach requires a `CLIENT_SECRET` be embedded in the binary, thus exposing the secret leaking by decompiling the binary
+- [Github Device Flow](https://docs.github.com/en/developers/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps#device-flow)
+  - This approach doesn't require a `CLIENT_SECRET` or callback page
+  - Probably won't work for a UI
+- [Gitlab Proof Key for Code Exchange (PKCE) Flow](https://docs.gitlab.com/ee/api/oauth2.html#authorization-code-with-proof-key-for-code-exchange-pkce)
+  - Note that the documentation explicitly calls out that PKCE is optimal for client-side apps without a public cloud server (our use case)
+  - The documentation example specifies a `CLIENT_SECRET`, but that may be a documentation bug. A `CLIENT_SECRET` should not be necessary
+- [Bitbucket Implicit Grant](https://developer.atlassian.com/cloud/bitbucket/oauth-2/#2--implicit-grant--4-2-)
+  - No `CLIENT_SECRET` required here
+
+### Kubernetes Cluster
+
+All operations will need to go through an API server that is installed on the cluster via the `wego gitops install` command. This API server will run with a service account that provides the ability to do the following:
+
+- Create `Namespace` resources
+- Create `Application`, `Source`, and `Kustomize|Helm` resources
+- Read all Kubernetes objects, in all namespaces
+
+### Security Concerns
+
+Each of the Git Providers listed above support some sort of "personal access token" that allow for "impersonation" style authorization. These tokens never expire, and provide "sudo" access to the user's Github account.
+
+For this reason, we want to utilize the more short-lived OAuth tokens to avoid exposing a personal access token. We also don't want to persist these tokens longer than we have to.
+
+In addition, flows that require a `CLIENT_SECRET` will require us to embed it in the binary, which leaves us open to exposing it via decompilation.
+
+### UX Concerns
+
+Given that the Weave GitOps UI and API server may be running on a user's cluster, we may not always know the OAuth callback URL ahead of time. For this reason, we will need to use OAuth flows that do not require a static callback URLs. The flows listed above in the Git Providers section fit that constraint.
+
+We will need to ensure that the callback URL is configurable by the user via environment variables. This will allow them to run and expose the UI/API servers on whichever endpoint they choose.
+
+## Design
+
+Since most modern Git Providers will support a form of browser-based OAuth, **we will utilize the browser for both the CLI and UI authentication with Git Providers**.
+
+In the case of the CLI, we can utilize a short-lived browser session that will recieve the OAuth callback and pass the token to the CLI:
```

Comment:
We don't need the local server running.  The CLI will act as the user agent to get the token.  Unfortunately, all 3 Git servers will require a different flow.

Reply:
How would the Git Provider redirect back to a CLI with the OAuth code?

Reply:
Jordan and I discussed this, the API server will be the URL the browser is redirected to.  When the browser calls this server, it will store the token which will then be retrieved from our CLI/UI.